crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3142,6 +3149,62 @@ impl<'db> TypeInferenceBuilder<'db> {
         )
     }
 
+    fn infer_binary_intersection_type_comparison(
+        &mut self,
+        intersection: IntersectionType<'db>,
+        op: ast::CmpOp,
+        other: Type<'db>,
+        intersection_on: IntersectionOn,
+    ) -> Result<Type<'db>, CompareUnsupportedError<'db>> {
+        // If a comparison yields a definitive true/false answer on a (positive) part of an
+        // intersection type, it will also yield a definitive answer on the full intersection
+        // type, which is even more specific.
+        for pos in intersection.positive(self.db) {
+            let result = match intersection_on {
+                IntersectionOn::Left => self.infer_binary_type_comparison(*pos, op, other)?,
+                IntersectionOn::Right => self.infer_binary_type_comparison(other, op, *pos)?,
+            };
+            if let Type::BooleanLiteral(b) = result {
+                return Ok(Type::BooleanLiteral(b));
+            }
+        }
+
+        for neg in intersection.negative(self.db) {
+            let result = match intersection_on {
+                IntersectionOn::Left => self.infer_binary_type_comparison(*neg, op, other)?,
+                IntersectionOn::Right => self.infer_binary_type_comparison(other, op, *neg)?,
+            };
+
+            match (op, result) {
+                (ast::CmpOp::Eq, Type::BooleanLiteral(true)) => {
+                    return Ok(Type::BooleanLiteral(false));
+                }
+                (ast::CmpOp::NotEq, Type::BooleanLiteral(false)) => {
+                    return Ok(Type::BooleanLiteral(true));
+                }
+                (ast::CmpOp::Is, Type::BooleanLiteral(true)) => {
+                    return Ok(Type::BooleanLiteral(false));
+                }
+                (ast::CmpOp::IsNot, Type::BooleanLiteral(false)) => {
+                    return Ok(Type::BooleanLiteral(true));
+                }
+                _ => {}
+            }
+        }
+
+        let simplified_type = if intersection.negative(self.db).is_empty() {
+            *intersection.positive(self.db).first().unwrap()
+        } else {
+            let mut builder = IntersectionBuilder::new(self.db);
+            for pos in intersection.positive(self.db) {
+                builder = builder.add_positive(*pos);
+            }
+            builder.build()
+        };
```

Comment:
A somewhat desperate attempt to get a reasonable answer if none of the special cases above have triggered. This is definitely not the final solution. Just something to prevent getting `Type::Todo` everywhere.